## Importing Libraries

In [7]:
# import libraries
import pandas as pd
import numpy as np
import os

## Importing Dataframe

In [8]:
## Importing data set path
path = r'C:\Users\mgril\OneDrive\Desktop\Instacart Basket Analysis Folder'

In [9]:
# Import ords_prods_merged
df_ords_prods_merged =  pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data','orders_products_merged.pkl'))

In [5]:
# Create subset of first 1M rows
df = df_ords_prods_merged[:1000000]

In [6]:
# Check shape
df.shape

(1000000, 14)

## Using IF statements

In [10]:
## define pricw label function 
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [11]:
# Apply price label on dataframe
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\mgril\AppData\Local\Temp\ipykernel_22880\2213225355.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [12]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [13]:
# Check max price to confirm
df['prices'].max()

14.8

## If-Statements with the loc() Function

In [15]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [16]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [17]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [18]:
# Check value counts
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [19]:
# Repeat process on entire dataframe
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, ' price_range_loc'] = 'High-range product'

In [20]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [21]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] < 5, 'price_range_loc'] = "Low-range product"

In [22]:
# Check value counts
df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    21860860
Low-range product     9900417
NaN                    643582
Name: count, dtype: int64

## If-Statements with For-Loops

In [23]:
# Create new column in dataframe
df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [24]:
# Set loop
result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [25]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [26]:
# Add result to dataframe
df_ords_prods_merged['busiest_day'] = result

In [27]:
# Check value counts
df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

## Task 02. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [31]:
# Create new if-statement for-loop for new requirements
result_v2 = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_v2.append("Busiest days")
  elif value == 4 or value == 3:
    result_v2.append("Least busy days")
  else:
    result_v2.append("Regularly busy")

In [32]:
# Add result to dataframe in new column
df_ords_prods_merged['busiest_days'] = result_v2

In [34]:
# Check
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  price_range_loc  \
0         Soda        77              7     9.0   both              NaN   
1         Soda        77              7     9.0   both              NaN   
2         Soda        77              7     9.0   both              NaN   
3         Soda        77              7     9.0   both              NaN   
4         Soda        77              7     9.0   both              NaN   

     price_range_loc     busiest_day     busiest_days  
0  Mid-range product  Regularly busy   Regularly busy  
1  Mid-range product  Regularly busy  Least busy days  
2  Mid-range product  Regularly busy  Least busy days  
3  Mid-range product      Least busy  Least busy days  
4  Mid-range product      Least busy  Least busy days

## Task 03. Check the values of this new column for accuracy. Note any observations in markdown format.

In [35]:
df_ords_prods_merged['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy     12916111
Busiest days       11864412
Least busy days     7624336
Name: count, dtype: int64

### 03. answer - Comparing with the original frequency counts, the output of values matches the groupings in sum for the total of the busiest days 0 and 1 = 11864412,  the values for least busiest days of 3 and 4 = 7624336, and regularly busy 2,5,6 = 12916111 confirming that the loop worked as intended. 

## Task 04. When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [36]:
# Check frequency of values in the "order_hour_of_day" column 
df_ords_prods_merged['order_hour_of_day'].value_counts()

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

### The client wants the above grouped instead of order hour of the day into periods of time broken into three parts. Break down into 24 hours of the day as follows: Most orders = 10, 11, 14, 15, 13, 12, 16, 9; Average orders = 17, 8, 18, 19, 20, 7, 21, 22; Fewest orders = 23, 6, 0, 1, 5, 2, 4, 3

In [38]:
# Create a for-loop if statement using the above breakdown
hour_of_day = []

for value in df_ords_prods_merged["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    hour_of_day.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    hour_of_day.append("Fewest orders")
  else:
    hour_of_day.append("Average orders")

In [39]:
# Add result to dataframe as "busiest_period_of_day" column
df_ords_prods_merged['busiest_period_of_day'] = hour_of_day

In [40]:
# Check
df_ords_prods_merged[['order_hour_of_day', 'busiest_period_of_day']].head(10)

order_hour_of_day busiest_period_of_day
0                  8        Average orders
1                  7        Average orders
2                 12           Most orders
3                  7        Average orders
4                 15           Most orders
5                  7        Average orders
6                  9           Most orders
7                 14           Most orders
8                 16           Most orders
9                  8        Average orders

## Task 05. Print the frequency for this new column.

In [41]:
df_ords_prods_merged['busiest_period_of_day'].value_counts()

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

### confirmed these sums match the frequencies of the ungrouped above in task 04. 

## Task 06. Ensure your notebook is clean and structured and that your code is well commented.

## Task 07. Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [42]:
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated.pkl'))